src/concurrency/thread.rs

```diff
@@ -671,6 +696,15 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
             }
         }
         self.yield_active_thread = false;
+
+        // The main thread is special, it is allowed to yield when it is shutting down and has
+        // nothing to execute. So if another thread yields back to the main thread when it is in
+        // this partly-shut-down state, we make another pass through the scheduler to either yield
+        // back off them main thread, or enter its shutdown sequence.
+        if self.active_thread == MAIN_THREAD && self.threads[self.active_thread].stack.is_empty() {
```

Comment:
Should this be inside the `if self.threads[self.active_thread].state == ThreadState::Enabled`? I think it should.

Reply:
Like this?
```diff
diff --git a/src/concurrency/thread.rs b/src/concurrency/thread.rs
index b59fcc8d..882c7eb7 100644
--- a/src/concurrency/thread.rs
+++ b/src/concurrency/thread.rs
@@ -684,19 +684,20 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
             debug_assert_ne!(self.active_thread, id);
             if thread.state == ThreadState::Enabled {
                 self.active_thread = id;
+                // The main thread is special, it is allowed to yield when it is shutting down and has
+                // nothing to execute. So if another thread yields back to the main thread when it is in
+                // this partly-shut-down state, we make another pass through the scheduler to either yield
+                // back off them main thread, or enter its shutdown sequence.
+                if self.active_thread == MAIN_THREAD
+                    && self.threads[self.active_thread].stack.is_empty()
+                {
+                    return self.schedule(clock);
+                }
                 break;
             }
         }
         self.yield_active_thread = false;
 
-        // The main thread is special, it is allowed to yield when it is shutting down and has
-        // nothing to execute. So if another thread yields back to the main thread when it is in
-        // this partly-shut-down state, we make another pass through the scheduler to either yield
-        // back off them main thread, or enter its shutdown sequence.
-        if self.active_thread == MAIN_THREAD && self.threads[self.active_thread].stack.is_empty() {
-            return self.schedule(clock);
-        }
-
         if self.threads[self.active_thread].state == ThreadState::Enabled {
             return Ok(SchedulingAction::ExecuteStep);
         }
```
This doesn't compile because `IndexVec::iter_enumerated` uses `impl Trait` in return position:
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> src/concurrency/thread.rs:694:28
    |
678 |           let threads = self
    |  _______________________-
679 | |             .threads
680 | |             .iter_enumerated()
    | |______________________________- immutable borrow occurs here
...
683 |           for (id, thread) in threads {
    |                               ------- a temporary with access to the immutable borrow is created here ...
...
694 |                       return self.schedule(clock);
    |                              ^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
...
698 |           }
    |           - ... and the immutable borrow might be used here, when that temporary is dropped and runs the destructor for type `std::iter::Chain<std::iter::Skip<impl DoubleEndedIterator + ExactSizeIterator + Iterator<Item = (concurrency::thread::ThreadId, &concurrency::thread::Thread<'_, '_>)> + '_>, std::iter::Take<impl DoubleEndedIterator + ExactSizeIterator + Iterator<Item = (concurrency::thread::ThreadId, &concurrency::thread::Thread<'_, '_>)> + '_>>`
```

Reply:
No i meant the one below. [Here](https://github.com/rust-lang/miri/blob/94ddca187760a1e139f8ac57d43f108268e74efd/src/concurrency/thread.rs#L674).